### DISCUSSION TITLE
Same command on IM 7 version yields different result than the command on 6 version

### DISCUSSION THREAD

**Discussion by eduardrudko** [ImageMagick/ImageMagick#7796] [Help] [answered]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

linux amd64 (Docker) on macOS Sonoma 14.6.1 (23G93)

### Description

Our application uses image magick 6 version on production, on linux amd64 platform. I'm upgrading to IM 7 as it seemed to have performance improvements over version 6, however i struggle to achieve same output TIF file on version 7 no matter how i craft my command. Let me give more context:
Here is -version output of both version
```bash
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
```
```bash
Version: ImageMagick 7.1.1-39 Q16 x86_64 22428 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(4.5)
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
Compiler: gcc (9.4)
```
This is the command we use to convert multi-page PDF file with embeded images to TIF file:
`convert -compress Group4 -density 400 -size 3400x4400 input.pdf output.tif`

I can't send the whole output TIF file, but i can show some fragments of it on a screenshot basically to get an idea of the problem. 
This is how embeded image from PDF converted to TIF file using the above command.
<img width="1680" alt="Screenshot 2024-11-20 at 19 47 14" src="https://github.com/user-attachments/assets/9eae5fbb-9c7b-4012-bfd3-47d58daa2f3e">

This is what i get using the same command but on version 7:

`/opt/bin/magick -compress Group4 -density 400 -size 3400x4400 input.pdf output.tif`

<img width="1677" alt="Screenshot 2024-11-20 at 19 50 37" src="https://github.com/user-attachments/assets/021b8f94-64f8-458a-b96b-004a1fa6d624">

Note: text look better.

Running `-verbose` option reveals that there is indeed some differencies:
Here i did the same but for the first page only to speed it up
IM6
```bash
ubuntu@e54c5c5754d4:/Testing$ convert -verbose -compress Group4 input.pdf[1] output.tif
'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72' -dFirstPage=2 -dLastPage=2 '-sOutputFile=magick-3227fZaCqOhYBhy%d' '-fmagick-322rY7zWCSoTJLx' '-fmagick-322mqEHPZjbEU8x'
magick-3227fZaCqOhYBhy1 PNG 595x842 595x842+0+0 8-bit sRGB 59461B 0.140u 0:00.199
input.pdf[1]=>input.pdf[1] PDF 595x842 595x842+0+0 16-bit sRGB 59461B 0.000u 0:00.000
input.pdf[1]=>output.tif[1] PDF 595x842 595x842+0+0 1-bit Bilevel Gray 2c 7520B 20.070u 1:25.009
```
IM7
```
ubuntu@e54c5c5754d4:/Testing$ /opt/bin/magick -verbose -compress Group4 input.pdf[1] output.tif
'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72' -dPrinted=false -dFirstPage=2 -dLastPage=2 '-sOutputFile=magick-VKFN13Bont5kicZU6UEKiPjTFQrObgV3%d' '-fmagick-mdEjAt6pKxetiRjGspdRTjNj1dtR46o5' '-fmagick-aCNNQWHlBxmqGp_aDvR4-uq_TWVh1qW5'
magick-VKFN13Bont5kicZU6UEKiPjTFQrObgV31 PNG 595x842 595x842+0+0 8-bit TrueColorAlpha sRGB 59461B 0.020u 0:00.345
input.pdf[1]=>input..pdf[1] PDF 595x842 595x842+0+0 16-bit TrueColorAlpha sRGB 59461B 0.000u 0:00.000
input.pdf[1]=>output.tif[1] PDF 595x842 595x842+0+0 1-bit Bilevel Gray 2c 5762B 0.220u 0:00.982
```

I can see there are difference here:
6 version: input.pdf[1]=>input.pdf[1] PDF 595x842 595x842+0+0 16-bit sRGB 59461B 0.000u 0:00.000
7 version: input.pdf[1]=>input..pdf[1] PDF 595x842 595x842+0+0 16-bit TrueColorAlpha sRGB 59461B 0.000u 0:00.000

but i frankly saying do not understand what that could mean, or what is the problem, however this may help to understand the problem.

Please advice, how can i achieve the same properties on version 7 of the tif image produced on 6 version?
The TIF file produced on version 6 has low file size (both compressed and decompressed) compressed is ~900 KBs, decompressed ~20MS
The TIF file produced on version 7 has poor quality and and very low file size (both compresed and decompressed) compressed is  ~64KBs using any of these FAX, Group4, uncompressed ~6 MB

My goal is to achieve similar look and file size with image magick 7 while benefiting from performance improvements. 
I can do some testing of my end and i am able to provide any additional debug data.
Thanks.

**Comment by fmw42** [comment]
IM 6.9.10.23 if I recall was buggy. Nevertheless

> convert -compress Group4 -density 400 -size 3400x4400 input.pdf output.tif

In that command -size does nothing, as I understand it. Use -resize if you want the result to be 3400x4400.

However, you should use the command in the right order. so

`convert -density 400 input.pdf -resize 3400x4400 -compress group4 output.tif`

In IM 7, then that would be

`magick -density 400 input.pdf -resize 3400x4400 -compress group4 output.tif`

And those should match. IM 7 is more strict about syntax order than was in IM 6. So read the input before specifying input processing and output arguments

**Comment by eduardrudko** [comment]
> In that command -size does nothing, as I understand it. Use -resize if you want the result to be 3400x4400.

@fmw42 thanks for a quick reaply. 
Correct. Empirically i found out that size has no effect in this scenario whatsoever, however i added it to doubled check before i reduce the scope of the problem.

Now, since we reduced the scope of the problem, lets entirely forget about reaching 3400x4400 and let it automatically calculate the end size.

Thus, running `/opt/bin/magick -density 400 input.pdf[1] -compress group4 output.tif` the problem still persist and the end result is identical to the image in the description. The images in TIF ile are distorted and almost black and it seem to be the entire image went though a totaly different algorithm(s) to produce the end image.

**Reply by fmw42** [reply]
Post the PDF file.

**Reply by eduardrudko** [reply]
@fmw42 
Here is a link to a dummy pdf file i've generated and was able to reproduce the problem.
input.pdf - https://drive.google.com/file/d/1ECDGQ8hMBZ0wJY7JKAy7xzjhuELJpU8x/view?usp=sharing

IM6 - `convert -compress Group4 -density 400 input.pdf output_im6_400.tif`
<img width="1434" alt="Screenshot 2024-11-21 at 16 28 23" src="https://github.com/user-attachments/assets/ca2f8e1c-81e3-4ed5-82bf-bfbbb0f6c507"> - **261 KBs**

IM7 - `/opt/bin/magick -density 400 input.pdf -compress group4 output_im7_400.tif`
<img width="1292" alt="Screenshot 2024-11-21 at 16 28 52" src="https://github.com/user-attachments/assets/86e58fc5-ae07-4499-bdb6-531e1f8fa6ed"> - **19 KBs**

My goal is achieve the same look and closest file size on IM7 as of the file produced on IM6

**Reply by dlemstra** [reply]
Attaching file for future reference: [input.zip](https://github.com/user-attachments/files/18049124/input.zip)

**Comment by fmw42** [comment]
I get exactly the same from 


`convert -density 400 input.pdf -compress group4 output_im6_400.tif`

`magick -density 400 input.pdf -compress group4 output_im7_400.tif`

[Archive.zip](https://github.com/user-attachments/files/17849169/Archive.zip)

```
magick compare -metric rmse output_im6_400.tif output_im7_400.tif null:
0 (0)
```

This is with  6.9.13-19 and 7.1.1-41

As I mentioned above, your syntax is wrong even for IM 6, though MAY be forgiving. Either of those old versions of IM might be buggy or have a change that was retracted later. I suggest that you upgrade.

**Comment by snibgo** [comment, marked as answer]
@eduardrudko wrote:
> IM6 - convert -compress Group4 -density 400 input.pdf output_im6_400.tif

... but the image shown is PNG, not a TIF. Why? Perhaps the PNG shown is a screenshot of a program that displays the TIF. This adds an unnecessary complication, and makes our task more difficult because we cannot directly compare our tests with the required result.

Nevertheless, the following makes a similar result to the claimed IM6 result (Windows syntax):
```
magick ^
  -density 400 input.pdf ^
  -colorspace Gray ^
  ( -size 1x1 xc:Black xc:White +append +repage -write mpr:MAP +delete ) ^
  -dither Floydsteinberg -remap mpr:MAP ^
  -compress Group4 x7.tiff
```
This command explicitly converts the image to grayscale and remaps with dithering to black and white.

Tested with v7.1.1-38 and v6.9.12-72, which give identical results:

**Comment by eduardrudko** [comment]
> ... but the image shown is PNG, not a TIF. Why? Perhaps the PNG shown is a screenshot of a program that displays the TIF. This adds an unnecessary complication, and makes our task more difficult because we cannot directly compare our tests with the required result.

You are right. These were the screenshots i made just to show the idea before i provide actual files. Here are the output files for version 6 and version 7 i get on my end.

6.9.10-23 Q16 x86_64 20190101 - https://drive.google.com/file/d/1UpONf8dOvO4X22mYxrHIzT0xBzeDc3Hh/view?usp=sharing

7.1.1-39 Q16 x86_64 22428  - https://drive.google.com/file/d/118OmcqzR3Iju-e8lEqURzrsJZTQRRGUc/view?usp=sharing

Your results are very similar to what i get on my version 6. This is strage. Version 6 achieves similar result with a simple input density change and output compression of group4, although version 7 needs that amount of adjustments. I guess this has to do with the channels optimisation that has happened in version 7 for grayscale images.

I'd like to understand how i can control the look of the image produced on version 7 to make it idential to what i have on version 6. 

@snibgo your command is very close to what i need, however it has no adjustable numeric values afar from density. My question is, is there anything that can be changed besides density to get a better look or perhaps at least the look i get on my version 6 produced?

**Reply by fmw42** [reply]
IM6 and IM 7 is correct (with proper syntax) on my machine with my version of IM.  It is possible that Group4 has changed with the changes to -monochrome.  But someone would have to check the code history. Have you tried the correct syntax on both IM 6 and IM 7? Your IM 7 is two version old, but not likely the issue. Your IM 6 is very old and should be upgraded.

**Reply by eduardrudko** [reply]
Yes, i did. On IM 6 using the correct syntax you have shown, which is:
`convert -density 400 input.pdf -compress group4 output_im6_400.tif`
produce the same result as the previous IM 6 command i have shown. Very slow execution, but good look and small size.
IM 7 i have already shown in the previous post. 

Indeed, is it even more strange that both Im6 and im7 on your end produce similar result that is equal to my result on im7 that is black distorted images. I'm not sure, but maybe group4 delegates something that is a different version. 

Also what is worth noting is that running 6 version has a big load on my CPU which is not the case with 7 version, here are some results i did several weeks ago:

```
IM v6
convert -verbose -limit thread 1  -density 600x600 input.pdf[1] -compress Group4 20241021-1-1jg2qqm.tiff
98% cpu 2:28,61 total

IM v7
magick -verbose -limit thread 1 -density 600x600 input.pdf[1] -compress Group4  20241021-1-1jg2qqm.tiff
61% cpu 0,490 total
```
this probably explains slowness on version 6, not the end look of the image though.

**Reply by fmw42** [reply]
As I understand it (possibly wrong), but group4 is binary. So it cannot look like the original input.pdf.  See https://en.wikipedia.org/wiki/Group_4_compression#:~:text=CCITT%20Group%204%20compression%2C%20also,%2Dand%2Dwhite)%20images.  In your old IM version, it might not have binarized, which could have been a bug or change to correct format.  

One of the IM developers or perhaps snibgo can confirm or comment

**Reply by eduardrudko** [reply]
I thought group4 does not change the look of the image since it is lossless but the IM itself prepares the image for the requirements of the compression algorithm, but maybe i'm just far fetching this thought without really knowing how it works.

**Comment by snibgo** [comment]
> I thought group4 does not change the look of the image since it is lossless ...

Well, that is not exactly correct. Group4 is a method of compression for pixels that are either black or white (aka "bilevel"). So if the image contains grays or colours, the image is first silently converted to bilevel, and this obviously changes the look of the image.

My command above explicitly changes the image to bilevel. I control the process, so I can decide how I want to do it, and the overall effect. For example, after converting to grayscale, I can darken the image ("-evaluate Pow 2") before remapping.
```
magick -density 400 input-1.pdf -colorspace Gray -evaluate Pow 2 ( -size 1x1 xc:Black xc:White +append +repage -write mpr:MAP +delete ) -dither Floydsteinberg -remap mpr:MAP -compress Group4 xp2.png
```
If you want to lighten instead of darken, use a power that is less than 1.0 (but more than 0.0).

**Comment by eduardrudko** [comment]
@snibgo Thanks, this was really insightful. Any thought why IM6 achieves this look by a mere `-density` option along with `-compression Group4` whereas IM7 require bunch of different options?

**Comment by snibgo** [comment]
> Any thought why IM6 achieves this look by ...

For me, using v6.9.12-72, I don't get the result you want. So I suppose there was in change between your v6.9.10-23 and my v6.9.12-72. I don't know what that change was.

**Reply by fmw42** [reply]
That is what I said above. Do you think it might be related to the -monochrome change? Otherwise it was a corrected bug or corrected feature.

**Reply by eduardrudko** [reply]
So trying to get to the bottom of this...is this how image compressed using group4 supposed to look like where embeded images look distorted and black with only showing a silhouette of the real image? And what is produced on my`v6.9.10-23` was decided to be incorrected behavior and it was fixed in between `v6.9.10-23` and `v6.9.12-72` to look this and it considered to be the better look or the "correct" look? I don't know sir but this image makes me to doubt it...
![af7fa4_76f1e6fda7454137a88e7b677993489f~mv1](https://github.com/user-attachments/assets/d502fa9d-c20d-4cc1-8a63-fabd0b9bc6f2) - [source](https://www.litigationsupporttipofthenight.com/single-post/2016/05/22/ccitt-group-4-tiff-images)

As the user of the tool i at least expect something to be one of the above look with a simple `-compression group4` option. Or maybe i'm just wrong and the black embeded images with a sole silhouette is indeed the the correct behavior? After all, it's been like that at least since `v6.9.12-72` and no one ever complained until now. 

Guess, that is the real question.
